### PR TITLE
Improve the validation of Monero addresses.

### PR DESCRIPTION
**Pull request by electrotype**
This code is based on existing implementations:
- https://github.com/bigreddmachine/MoneroPy/blob/master/moneropy/base58.py
- https://github.com/ognus/wallet-address-validator/blob/master/src/monero_validator.js

More doc: 
- https://monerodocs.org/cryptography/base58/

The PR idea comes from: 
- https://monero.stackexchange.com/questions/11626/monero-address-validation-using-java

----------

Please note that I'm not an expert at this level of coding (dealing with byte[], encryption, etc.). I tried my best to convert the code from implementations in other languages to Java. But maybe I use `int[]` at some places where `byte[]` should be used and/or the opposite!

Feel free to improve the code!

Also:
- I wasn't 100% sure where to put the code in the project.
- I'm not sure about the rules for the STAGENET network. I don't think I've seen any tests for it anywhere... This PR currently uses the same code as the TESTNET network.
- I had to add `Bouncy Castle` as a dependency for `Keccak 256` digests.
- I only ran the new tests (for this PR), since I currently do not have any Monero deamon running. Maybe I broke some other tests. Please check!

### CODE REVIEW

File: src/main/java/common/utils/GenUtils.java [hunk 1]
```diff
@@ -44,4 +46,30 @@ public static String getIndent(int length) {
     for (int i = 0; i < length; i++) str += "  "; // two spaces
     return str;
   }
+  
+  /**
+   * Produces a new array containing the elements between
+   * the start and end indices.
+   * <p>
```
Comment (OlivierAlbertini) [inline src/main/java/common/utils/GenUtils.java]: missing `</p>`
and `@param` descriptions like you did for `validateAddress`

File: src/main/java/monero/utils/MoneroUtils.java [hunk 2]
```diff
@@ -82,10 +114,71 @@ public static void validatePublicSpendKey(String publicSpendKey) {
...
-  // TODO: improve validation
+  /**
+   * Validate an address but ignore the type of network.
+   */
   public static void validateAddress(String address) {
+    validateAddress(address, null);
+  }
+  
+  /**
+   * @param moneroNetworkType if <code>null</code> the validation will
```
Comment (OlivierAlbertini) [inline src/main/java/monero/utils/MoneroUtils.java]: missing `@param address`

File: src/test/java/test/TestMoneroWalletGeneric.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.junit.Test;
+
+import monero.daemon.model.MoneroNetworkType;
+import monero.utils.MoneroUtils;
+
+public class TestMoneroWalletGeneric {
+
+  @Test
+  public void testAddressValidation() {
```
Comment (woodser) [inline src/test/java/test/TestMoneroWalletGeneric.java]: Tests must cover STAGENET.  Network type should either be specified or default to MAINNET.

**Comment by woodser** [comment]
@electrotype 

> I'm not sure about the rules for the STAGENET network. I don't think I've seen any tests for it anywhere... 

The JUnit wallet and daemon tests will test whichever network is configured in TestUtils.java.  In this case, these are static utilities which do not use a wallet instance, so testing the static utilities within `TestMoneroUtils.java` with coverage for the 3 network types seems appropriate.

File: src/test/java/test/TestMoneroWalletGeneric.java (hunk 3 above)
Comment (woodser) [reply src/test/java/test/TestMoneroWalletGeneric.java]: These tests should show that invalid addresses, including mixing up network type, will throw an exception.

**Comment by electrotype** [comment]
@woodser Thanks for your review. As I said, do what you want with this code. I did try to adjust it a bit from my own project to yours but I do not have time for more. I just wanted to share it, since I didn't find any examples in Java. 

Feel free to copy/keep any part of this PR you want and delete the PR afterward. I don't care about credits...

I'll try to find time to contribute more in the future. Thanks!

**Comment by woodser** [comment]
This is still very helpful.  I will try to incorporate it while preserving your commit history soon.  Thank you again @electrotype!

**Pull request merged**